### PR TITLE
Set scheme auto-detection to disfavor example and demo schemes

### PR DESCRIPTION
**Pull request by yhkaplan**
## What
I made the scheme auto-detection avoid setting the scheme to names with "E/example" or "D/demo" in their name.

## Why
I found it a little tedious to set a custom vimrc for every project.

## How
I added a line to `list_schemes.sh` to exclude the scheme names above.

## Questions/Concerns
Any other scheme names I should disfavor?

### CODE REVIEW

**Comment by gfontenot** [comment]
This is a great idea! Thanks!

I think that the only change I'd like to see is for this to be configurable, and for it to default to not filtering anything. I think we could accomplish this by adding a new `g:xcode_scheme_ignore_pattern` and passing it through to the `list_schemes.sh` script. If we don't have anything set for an ignore pattern, we shouldn't filter anything out.

**Comment by yhkaplan** [comment]
Thanks for the suggestion! I don't have too much experience with Vim-script, but I'll try implementing it.

**Comment by yhkaplan** [comment]
I think I got the vim-script part working. I just have to fix list_schemes.sh to switch to non-ignore mode when no arguments are provided and test it out more.

**Comment by yhkaplan** [comment]
It seems to be working now, but using `if [[ "$ignore_pattern" == "-workspace" ]]` feels kinda hacky...

File: bin/list_schemes.sh [hunk 1]
```diff
@@ -2,7 +2,17 @@
...
 set -o pipefail
 
-xcrun xcodebuild -list "$@" 2>/dev/null \
-  | awk '/Schemes:/,0' \
-  | tail -n +2 \
-  | sed -e "s/^[[:space:]]*//"
+ignore_pattern="$1"
+
+if [[ "$ignore_pattern" == "-workspace" ]]; then
+    xcrun xcodebuild -list 2>/dev/null \
```
Comment (gfontenot) [inline bin/list_schemes.sh]: I think these are losing the reference to the build target now, right? We need to keep passing this through to make sure we're listing the schemes in the selected project/workspace, as opposed to the first one `xcodebuild` finds.

**Review by gfontenot — COMMENTED** [review]
This is a great start! As noted, we need to make sure that we keep our reference to the project/workspace when calling `xcodebuild -list`, and I agree that checking for `-workspace` probably isn't the best solution for switching the behavior.

I might suggest using command line flags here in order to clarify our behavior. They are... non-obvious in shell, but basically we can use `getopts` to provide a trivial interface:

```shell
while getopts "t:i:" opt; do
    case $opt in
        t) target="$OPTARG";;
        i) ignore_pattern="$OPTARG";;
        \?)
            echo "Invalid option: -$OPTARG" >&2
            exit 1
            ;;
        :)
            echo "Option -$OPTARG requires an argument." >&2
            exit 1
            ;;
    esac
done
```

This will give us access to the pattern and the target directly as named variables. The vimscript code will just need to add `"-i"` to the command line args for the case where we have an ignore pattern, and we'll need to add `"-t"` to the target as well. The interface is clunky, but since this is an internal tool it should be plenty for our needs.

Also, instead of duplicating the code to list the schemes, I think we should maybe list the schemes the same way every time, cached into a temp variable, and then filter out the ignore pattern if it exists. This will make the whole thing more extensible later on as other needs arise.

Does this sound reasonable to you? I'm happy to help out with this, I think this is all a super good idea and I'm totally on board with it.

**Comment by yhkaplan** [comment]
That sounds like a much better approach, thanks! I’m excited to work on it! I should have time this weekend.

**Comment by yhkaplan** [comment]
I think I've got it working, just gotta test it out a little more.

**Comment by yhkaplan** [comment]
@gfontenot I'm not sure what approach I should use to combine the two strings "-workspace" and "{workspace-name}" together, then reformat them to be passed into `xcodebuild -list`. There would need to be a space between the two strings, so one approach would be to replace the space with a delimiter, and then change the delimiter back into a space in the shell script, but that would appear to require more changes to the vim script part, which seems a little high maintenance. 

Getopts only seems to pick up the "-workspace" half,  but do you have any advice about how to overcome this?

**Comment by gfontenot** [comment]
Hey @yhkaplan, sorry for the delayed response. Busy weekend. I need to think about this a bit more. It's an interesting problem because we need to make sure that the arguments passed to `-workspace` retain their quotes as well.

Just spitballing, but maybe instead of passing the entire workspace command as one piece we split it up so that we're passing `-p -workspace -t "Path/To My/Workspace"` or something? Seems wonky, but it might be the best we can do?

**Comment by yhkaplan** [comment]
Thanks for the advice! I initially thought splitting them would make xcode.vim more complicated, but it didn’t seem to have too big an impact, so I pushed the commits! What do you think?

File: bin/list_schemes.sh [hunk 2]
```diff
@@ -2,7 +2,32 @@
 
 set -o pipefail
 
-xcrun xcodebuild -list "$@" 2>/dev/null \
+while getopts "f:t:i:" opt; do
+  case $opt in
+    f) flag="$OPTARG";;
```
Comment (gfontenot) [inline bin/list_schemes.sh]: Nit: is there a more descriptive name here other than `flag`? Maybe `target_type_flag`? Not super important, I can just see myself being confused by this at some point in the future.

**Review by gfontenot — COMMENTED** [review]
This is _so good_! Two more small requests if you feel like taking them on, if not I can handle them after merging:

 - Can you run `shellcheck` (`brew install shellcheck`) against `list_schemes.sh` to check for shell validity?
 - Can you add `g:xcode_scheme_ignore_pattern` to the help file and the README?

Thanks for your work on this! I'm pumped with how it's turned out.

File: bin/list_schemes.sh (hunk 2 above)
Comment (yhkaplan) [reply bin/list_schemes.sh]: Done here!: ded7e553708d5d4bcfe56a50c95efc17bb40cc6b

**Comment by yhkaplan** [comment]
I ran shellcheck and made additions to the help file and README

**Comment by gfontenot** [comment]
🎉 Awesome work! Thanks for your patience putting it together! I squashed and merged this as 2adc409, so it is now available to use on master.

**Comment by yhkaplan** [comment]
Thanks for all the guidance and support!

**Pull request closed without merging**